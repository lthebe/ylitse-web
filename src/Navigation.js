import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import AppBar from 'material-ui/AppBar';
import Button from 'material-ui/Button';
import Menu, { MenuItem } from 'material-ui/Menu';
import Snackbar from 'material-ui/Snackbar';
import { withStyles } from 'material-ui/styles';
import Toolbar from 'material-ui/Toolbar';
import Typography from 'material-ui/Typography';
import IconButton from 'material-ui/IconButton';
import CloseIcon from 'material-ui-icons/Close';
import InfoIcon from 'material-ui-icons/Info';
import MoreVertIcon from 'material-ui-icons/MoreVert';

import { version as uiVersion } from '../package.json';
import AboutDialog from './AboutDialog';

const styles = () => ({
    flex: {
        flex: 1,
    },
    iconButton: {
        marginLeft: 20,
        marginRight: -24,
    },
});

const AccountsLink = props => <Link to="/accounts" {...props} />;
const SkillsLink = props => <Link to="/skills" {...props} />;

class Navigation extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            flex: PropTypes.string,
            infoButton: PropTypes.string,
        }).isRequired,
    }

    constructor(props) {
        super(props);

        this.state = {
            aboutOpen: false,
            apiVersion: 'N/A',
            menuAnchor: null,
            errorOpen: false,
            errorMessage: '',
        };
    }

    componentDidMount() {
        this.fetchVersion();
    }

    fetchVersion = async () => {
        try {
            const resp = await fetch('/api/version', {
                credentials: 'include',
                redirect: 'follow',
            });
            if (resp.redirected) {
                window.location.replace(resp.url);
            }
            const data = await resp.json();

            this.setState({ apiVersion: data.api });
        } catch (e) {
            this.openError(e.message);
        }
    }

    logout = () => {
        window.location.replace('/logout');
    }

    openAbout = () => {
        this.setState({ aboutOpen: true });
    }

    closeAbout = () => {
        this.setState({ aboutOpen: false });
    }

    openMenu = (event) => {
        this.setState({ menuAnchor: event.currentTarget });
    };

    closeMenu = () => {
        this.setState({ menuAnchor: null });
    };

    openError = (errorMessage) => {
        this.setState({ errorMessage, errorOpen: true });
    }

    closeError = () => {
        this.setState({ errorOpen: false });
    }

    render() {
        const { menuAnchor } = this.state;
        const { classes } = this.props;

        return (
            <AppBar position="static">
                <Toolbar display="flex">
                    <Typography
                        variant="title"
                        color="inherit"
                        className={classes.flex}
                    >
                        Ylitse Admin
                    </Typography>
                    <Button component={AccountsLink} color="inherit">
                        Accounts
                    </Button>
                    <Button component={SkillsLink} color="inherit">
                        Skills
                    </Button>
                    <IconButton
                        color="inherit"
                        className={classes.iconButton}
                        onClick={this.openAbout}
                    >
                        <InfoIcon />
                    </IconButton>
                    <IconButton
                        aria-owns={menuAnchor ? 'menu-appbar' : null}
                        aria-haspopup="true"
                        color="inherit"
                        className={classes.iconButton}
                        onClick={this.openMenu}
                    >
                        <MoreVertIcon />
                    </IconButton>
                </Toolbar>
                <Menu
                    open={Boolean(menuAnchor)}
                    id="menu-appbar"
                    anchorEl={menuAnchor}
                    onClose={this.closeMenu}
                >
                    <MenuItem onClick={this.logout}>Sign out</MenuItem>
                </Menu>
                <AboutDialog
                    open={this.state.aboutOpen}
                    uiVersion={uiVersion}
                    apiVersion={this.state.apiVersion}
                    onOkClick={this.closeAbout}
                    onClose={this.closeAbout}
                />
                <Snackbar
                    open={this.state.errorOpen}
                    message={this.state.errorMessage}
                    autoHideDuration={3000}
                    action={
                        <IconButton color="inherit" onClick={this.closeError}>
                            <CloseIcon />
                        </IconButton>
                    }
                    onClose={this.closeError}
                />
            </AppBar>
        );
    }
}

export default withStyles(styles)(Navigation);
