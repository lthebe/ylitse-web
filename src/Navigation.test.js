import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow, mount } from 'enzyme';

import Navigation from './Navigation';

describe('Navigation', () => {
    const navigation = <MemoryRouter><Navigation /></MemoryRouter>;

    test('renders correctly', () => {
        const wrapper = shallow(navigation).dive().dive().shallow();

        expect(wrapper).toMatchSnapshot();
    });

    test('has the right title', () => {
        const wrapper = shallow(navigation).dive().dive().shallow();
        const title = 'Ylitse Admin';

        expect(wrapper.find('[type="title"]').render().text()).toBe(title);
    });

    test('about dialog is hidden', () => {
        const wrapper = shallow(navigation).dive().dive().shallow();

        expect(wrapper.find('AboutDialog').prop('open')).toBe(false);
    });

    test('about dialog can be opened and closed', () => {
        const wrapper = mount(navigation);

        wrapper.find('Toolbar').find('IconButton').simulate('click');
        expect(wrapper.find('AboutDialog').prop('open')).toBe(true);
        wrapper.find('AboutDialog').find('Button').simulate('click');
        expect(wrapper.find('AboutDialog').prop('open')).toBe(false);
    });

    test('fetched version is shown in about dialog', async () => {
        const wrapper = shallow(navigation).dive().dive().shallow();
        const version = '0.1.0';
        const res = new Response(`{"api": "${version}"}`, {
            status: 200,
            headers: {
                'content-type': 'application/json',
            },
        });

        window.fetch = jest.fn();
        window.fetch.mockReturnValue(Promise.resolve(res));

        await wrapper.instance().fetchVersion();
        wrapper.update();
        expect(wrapper.find('AboutDialog').prop('apiVersion')).toBe(version);
    });

    test('error message is shown upon fetch error', async () => {
        const wrapper = shallow(navigation).dive().dive().shallow();
        const error = 'Test error';

        window.fetch = jest.fn();
        window.fetch.mockReturnValue({
            json: () => {
                throw new Error(error);
            },
        });

        await wrapper.instance().fetchVersion();
        wrapper.update();
        expect(wrapper.find({ message: error, open: true })).toHaveLength(1);
        wrapper.instance().closeError();
        wrapper.update();
        expect(wrapper.find({ message: error, open: false })).toHaveLength(1);
    });
});
