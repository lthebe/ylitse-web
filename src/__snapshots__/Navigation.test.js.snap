// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navigation
    classes={
      Object {
        "flex": "Navigation-flex-1",
        "iconButton": "Navigation-iconButton-2",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(Toolbar)
          display="flex"
        >
          <WithStyles(Typography)
            className="Navigation-flex-1"
            color="inherit"
            variant="title"
          >
            Ylitse Admin
          </WithStyles(Typography)>
          <WithStyles(Button)
            color="inherit"
            component={[Function]}
          >
            Accounts
          </WithStyles(Button)>
          <WithStyles(Button)
            color="inherit"
            component={[Function]}
          >
            Skills
          </WithStyles(Button)>
          <WithStyles(IconButton)
            className="Navigation-iconButton-2"
            color="inherit"
            onClick={[Function]}
          >
            <pure(Info) />
          </WithStyles(IconButton)>
          <WithStyles(IconButton)
            aria-haspopup="true"
            aria-owns={null}
            className="Navigation-iconButton-2"
            color="inherit"
            onClick={[Function]}
          >
            <pure(MoreVert) />
          </WithStyles(IconButton)>
        </WithStyles(Toolbar)>,
        <WithStyles(Menu)
          anchorEl={null}
          id="menu-appbar"
          onClose={[Function]}
          open={false}
        >
          <WithStyles(MenuItem)
            onClick={[Function]}
          >
            Sign out
          </WithStyles(MenuItem)>
        </WithStyles(Menu)>,
        <AboutDialog
          apiVersion="N/A"
          onClose={[Function]}
          onOkClick={[Function]}
          open={false}
          uiVersion="0.1.0+git"
        />,
        <WithStyles(Snackbar)
          action={
            <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(Close) />
            </WithStyles(IconButton)>
          }
          autoHideDuration={3000}
          message=""
          onClose={[Function]}
          open={false}
        />,
      ],
      "position": "static",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Typography)
              className="Navigation-flex-1"
              color="inherit"
              variant="title"
            >
              Ylitse Admin
            </WithStyles(Typography)>,
            <WithStyles(Button)
              color="inherit"
              component={[Function]}
            >
              Accounts
            </WithStyles(Button)>,
            <WithStyles(Button)
              color="inherit"
              component={[Function]}
            >
              Skills
            </WithStyles(Button)>,
            <WithStyles(IconButton)
              className="Navigation-iconButton-2"
              color="inherit"
              onClick={[Function]}
            >
              <pure(Info) />
            </WithStyles(IconButton)>,
            <WithStyles(IconButton)
              aria-haspopup="true"
              aria-owns={null}
              className="Navigation-iconButton-2"
              color="inherit"
              onClick={[Function]}
            >
              <pure(MoreVert) />
            </WithStyles(IconButton)>,
          ],
          "display": "flex",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Ylitse Admin",
              "className": "Navigation-flex-1",
              "color": "inherit",
              "variant": "title",
            },
            "ref": null,
            "rendered": "Ylitse Admin",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Accounts",
              "color": "inherit",
              "component": [Function],
            },
            "ref": null,
            "rendered": "Accounts",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Skills",
              "color": "inherit",
              "component": [Function],
            },
            "ref": null,
            "rendered": "Skills",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <pure(Info) />,
              "className": "Navigation-iconButton-2",
              "color": "inherit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "aria-haspopup": "true",
              "aria-owns": null,
              "children": <pure(MoreVert) />,
              "className": "Navigation-iconButton-2",
              "color": "inherit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "anchorEl": null,
          "children": <WithStyles(MenuItem)
            onClick={[Function]}
          >
            Sign out
          </WithStyles(MenuItem)>,
          "id": "menu-appbar",
          "onClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Sign out",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Sign out",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "apiVersion": "N/A",
          "onClose": [Function],
          "onOkClick": [Function],
          "open": false,
          "uiVersion": "0.1.0+git",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": <WithStyles(IconButton)
            color="inherit"
            onClick={[Function]}
          >
            <pure(Close) />
          </WithStyles(IconButton)>,
          "autoHideDuration": 3000,
          "message": "",
          "onClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Toolbar)
            display="flex"
          >
            <WithStyles(Typography)
              className="Navigation-flex-1"
              color="inherit"
              variant="title"
            >
              Ylitse Admin
            </WithStyles(Typography)>
            <WithStyles(Button)
              color="inherit"
              component={[Function]}
            >
              Accounts
            </WithStyles(Button)>
            <WithStyles(Button)
              color="inherit"
              component={[Function]}
            >
              Skills
            </WithStyles(Button)>
            <WithStyles(IconButton)
              className="Navigation-iconButton-2"
              color="inherit"
              onClick={[Function]}
            >
              <pure(Info) />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              aria-haspopup="true"
              aria-owns={null}
              className="Navigation-iconButton-2"
              color="inherit"
              onClick={[Function]}
            >
              <pure(MoreVert) />
            </WithStyles(IconButton)>
          </WithStyles(Toolbar)>,
          <WithStyles(Menu)
            anchorEl={null}
            id="menu-appbar"
            onClose={[Function]}
            open={false}
          >
            <WithStyles(MenuItem)
              onClick={[Function]}
            >
              Sign out
            </WithStyles(MenuItem)>
          </WithStyles(Menu)>,
          <AboutDialog
            apiVersion="N/A"
            onClose={[Function]}
            onOkClick={[Function]}
            open={false}
            uiVersion="0.1.0+git"
          />,
          <WithStyles(Snackbar)
            action={
              <WithStyles(IconButton)
                color="inherit"
                onClick={[Function]}
              >
                <pure(Close) />
              </WithStyles(IconButton)>
            }
            autoHideDuration={3000}
            message=""
            onClose={[Function]}
            open={false}
          />,
        ],
        "position": "static",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(Typography)
                className="Navigation-flex-1"
                color="inherit"
                variant="title"
              >
                Ylitse Admin
              </WithStyles(Typography)>,
              <WithStyles(Button)
                color="inherit"
                component={[Function]}
              >
                Accounts
              </WithStyles(Button)>,
              <WithStyles(Button)
                color="inherit"
                component={[Function]}
              >
                Skills
              </WithStyles(Button)>,
              <WithStyles(IconButton)
                className="Navigation-iconButton-2"
                color="inherit"
                onClick={[Function]}
              >
                <pure(Info) />
              </WithStyles(IconButton)>,
              <WithStyles(IconButton)
                aria-haspopup="true"
                aria-owns={null}
                className="Navigation-iconButton-2"
                color="inherit"
                onClick={[Function]}
              >
                <pure(MoreVert) />
              </WithStyles(IconButton)>,
            ],
            "display": "flex",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Ylitse Admin",
                "className": "Navigation-flex-1",
                "color": "inherit",
                "variant": "title",
              },
              "ref": null,
              "rendered": "Ylitse Admin",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Accounts",
                "color": "inherit",
                "component": [Function],
              },
              "ref": null,
              "rendered": "Accounts",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Skills",
                "color": "inherit",
                "component": [Function],
              },
              "ref": null,
              "rendered": "Skills",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <pure(Info) />,
                "className": "Navigation-iconButton-2",
                "color": "inherit",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "aria-haspopup": "true",
                "aria-owns": null,
                "children": <pure(MoreVert) />,
                "className": "Navigation-iconButton-2",
                "color": "inherit",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "anchorEl": null,
            "children": <WithStyles(MenuItem)
              onClick={[Function]}
            >
              Sign out
            </WithStyles(MenuItem)>,
            "id": "menu-appbar",
            "onClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Sign out",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Sign out",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "apiVersion": "N/A",
            "onClose": [Function],
            "onOkClick": [Function],
            "open": false,
            "uiVersion": "0.1.0+git",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(Close) />
            </WithStyles(IconButton)>,
            "autoHideDuration": 3000,
            "message": "",
            "onClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
